sys/aarch64/board.c

```diff
@@ -76,16 +119,31 @@ static void rpi3_physmem(void) {
   paddr_t kern_end = (paddr_t)_bootmem_end;
   paddr_t rd_start = ramdisk_get_start();
   paddr_t rd_end = rd_start + ramdisk_get_size();
+  paddr_t dtb_start = dtb_early_root();
+  paddr_t dtb_end = dtb_start + fdt_totalsize(PHYS_TO_DMAP(dtb_start));
+
+  addr_range_t memory[3] = {
+    {.start = rounddown(kern_start, PAGESIZE),
+     .end = roundup(kern_end, PAGESIZE)},
+    {.start = rounddown(rd_start, PAGESIZE), .end = roundup(rd_end, PAGESIZE)},
+    {.start = rounddown(dtb_start, PAGESIZE),
+     .end = roundup(dtb_end, PAGESIZE)},
+  };
 
-  vm_physseg_plug(ram_start, kern_start);
-  vm_physseg_plug_used(kern_start, kern_end);
+  qsort(memory, 3, sizeof(addr_range_t), addr_range_cmp);
 
-  if (rd_start != rd_end) {
-    vm_physseg_plug(kern_end, rd_start);
-    vm_physseg_plug_used(rd_start, rd_end);
-    vm_physseg_plug(rd_end, ram_end);
-  } else {
-    vm_physseg_plug(kern_end, ram_end);
+  vm_physseg_plug(ram_start, memory[0].start);
```

Comment:
This at least calls for FIXME annotation. It'd be better if `vm_physseg_plug*` could handle our use cases without such hacks. If that's not fixed in this PR then I'd at least expect a GitHub issue to be created for it. in fact that could be a great task for beginners.

Reply:
It's not a place for these changes.
I have added comment and created a new issue.